pkg/health/health.go

```diff
@@ -0,0 +1,160 @@
+// determined and Diagnosis.EnoughRequests is set to false.
+//
+// Otherwise, all metrics criteria are checked to determine if the revision is
+// healthy.
+func Diagnose(ctx context.Context, provider metrics.Metrics, query metrics.Query,
+	offset time.Duration, minRequests int64, healthCriteria []config.Metric) (*Diagnosis, error) {
+
+	logger := util.LoggerFromContext(ctx)
+	metricsValues, err := CollectMetrics(ctx, provider, query, offset, healthCriteria)
+	if err != nil {
+		return nil, errors.Wrap(err, "could not collect metrics")
+	}
+
+	overallResult := Healthy
+	var results []*CheckResult
+	for i, criteria := range healthCriteria {
+		result := determineResult(criteria.Type, criteria.Threshold, metricsValues[i])
+		results = append(results, result)
+
+		if !result.IsCriteriaMet {
+			overallResult = Unhealthy
+
+			logger := logger.WithFields(logrus.Fields{
+				"metricsType": criteria.Type,
+				"threshold":   criteria.Threshold,
+				"actualValue": result.ActualValue,
+			})
+			// If criteria is a latency, we want to log the percentile as well
+			if criteria.Type == config.LatencyMetricsCheck {
+				logger = logger.WithField("percentile", criteria.Percentile)
+			}
+			logger.Debug("criteria was not met")
+		}
+	}
+
+	return &Diagnosis{
+		OverallResult: overallResult,
+		CheckResults:  results,
+	}, nil
+}
+
+// CollectMetrics returns an array of values collected for each of the specified
+// metrics criteria.
+func CollectMetrics(ctx context.Context, provider metrics.Metrics, query metrics.Query, offset time.Duration, healthCriteria []config.Metric) ([]float64, error) {
+	logger := util.LoggerFromContext(ctx)
+	logger.Debug("start collecting metrics")
+	var values []float64
+	for _, criteria := range healthCriteria {
+		var value float64
+		var err error
+
+		switch criteria.Type {
+		case config.LatencyMetricsCheck:
+			value, err = latency(ctx, provider, query, offset, criteria.Percentile)
+			break
+		case config.ErrorRateMetricsCheck:
+			value, err = errorRatePercent(ctx, provider, query, offset)
+			break
+		default:
+			return nil, errors.Errorf("unimplemented metrics %q", criteria.Type)
+		}
+
+		if err != nil {
+			return nil, errors.Wrapf(err, "failed to obtain metrics %q", criteria.Type)
+		}
+		values = append(values, value)
+	}
+
+	return values, nil
+}
+
+// determineResult concludes if metrics criteria was met.
+//
+// The returned value also includes a string with details of why the criteria
+// was met or not.
+func determineResult(metricsType config.MetricsCheck, threshold float64, actualValue float64) *CheckResult {
+	result := &CheckResult{ActualValue: actualValue, Threshold: threshold}
+
+	// As of now, the supported health criteria (latency and error rate) need to
+	// be less than the threshold. So, this is sufficient for now but might need
+	// to change to a switch statement when criteria with a minimum threshold is
+	// added.
+	if actualValue <= threshold {
+		result.IsCriteriaMet = true
```

Comment:
@grayside so it's actually hard to look at `CheckResult` object and tell if it's a positive result or negative. 

e.g. for latency metrics, threshold=100 actual=200 → requirement is **failed**
however, for "min request" metrics threshold=100 actual=200 → requirement is **met**

the method that generates this CheckResult decides if those numbers mean something good or bad.

in case of "min request" metrics is **not met** → we simply do noop (wait more requests to happen)

however in case of "latency" metrics **not met** → means "rollback".

Reply:
Ah, that makes sense, and reading again I see the comment about the forthcoming switch statement. That seems like it will be a bit complex. I can't help but go to the idea that there should be an interface such as the following (reusing the same verbs being discussed above, but more locally this time):

```go
type Metric interface {
  Assess(context.Context, metrics.Metrics, time.Duration, []config.Metrics[]) (*CheckResult, error)
  Diagnose() bool
}
```

Then the individual Metrics can be implemented with a bit more isolation and testability.
The current structure seems like it will work okay for 2-3 metrics but start to become less manageable thereafter.

Reply:
Thanks @grayside! I think I understand your concern, but I am a little bit confused about your expected interface for the caller of this interface.

The idea is that you get a `CheckResult` per metric criteria since the `CheckResult` contains the expected threshold, the actual value, and whether the criteria is met based on those previous two values.

1) Do you want to pass a single `config.Metric` and get a single `CheckResult`?
```
for criteria := range healthCriteria {
    results := append(results, Assess(ctx, metricsProvider, ....)) // ignore error
}
```

2) Do you want to pass an array of metrics criteria an get an array of `CheckResult`?

3) Something else?

Also, I don't see how this will prevent the `switch` when determining results since using comparing `actualValue <= threshold` or `threshold <= actualValue` depends on the metrics criteria type (unless we pass a comparison function, which sounds like an overkill here).

Reply:
I'm a bit +1 on this being an overkill right now.
Realistically we'll only have `a≤b` or `b≤a` comparisons as far as metrics go.
So creating a `LatencyMetric` or `RequestCountMetric` types probably are overkill.

Though I agree, Adam’s approach long-term is definitely appealing. I'd say no need to refactor this right now.